adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m

```diff
@@ -0,0 +1,220 @@
+            return nil;
+        }
+        id<GADMediationNativeAdEventDelegate> delegate = nil;
+        if (originalCompletionHandler) {
+            delegate = originalCompletionHandler(ad, error);
+        }
+        originalCompletionHandler = nil;
+        return delegate;
+    };
+    
+    NSString *placementId = adConfiguration.credentials.settings[GADMAdapterPanglePlacementID] ?: @"";
+    if (!placementId.length) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInvalidServerParameters,
+                                                                      [NSString stringWithFormat:@"%@ cannot be nil.",GADMAdapterPanglePlacementID]
+                                                                      );
+        _loadCompletionHandler(nil, error);
+        return;
+    }
+    
+    BUAdSlot *slot = [[BUAdSlot alloc] init];
+    slot.ID = placementId;
+    slot.AdType = BUAdSlotAdTypeFeed;
+    
+    _nativeAd = [[BUNativeAd alloc]initWithSlot:slot];
+    if (!_nativeAd) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInitAd,
+                                                                      @"Init ad error, Please check the reason or you could contact Pangle"
+                                                                      );
+        _loadCompletionHandler(nil, error);
+        return;
+    }
+    _nativeAd.delegate = self;
+    [_nativeAd setAdMarkup:adConfiguration.bidResponse];
+}
+
+- (BUNativeAdRelatedView *)getRelatedView {
+    if (!_relatedView) {
+        _relatedView = [[BUNativeAdRelatedView alloc] init];
+    }
+    return _relatedView;
+}
+
+#pragma mark GADMediationNativeAd
+
+- (UIView *)mediaView {
+  if (!_mediaView) {
+    BUVideoAdView *videoView = [self getRelatedView].videoAdView;
+    _mediaView = videoView;
+  }
+  return _mediaView;
+}
+
+- (UIView *)adChoicesView {
+    return [self getRelatedView].logoADImageView;
+}
+
+- (NSString *)headline {
+    if (_nativeAd && _nativeAd.data) {
+        return _nativeAd.data.AdTitle;
+    }
+    return nil;
+}
+
+- (NSString *)body {
+    if (_nativeAd && _nativeAd.data) {
+        return _nativeAd.data.AdDescription;
+    }
+    return nil;
+}
+
+- (NSString *)callToAction {
+    if (_nativeAd && _nativeAd.data) {
+        return _nativeAd.data.buttonText;
+    }
+    return nil;
+}
+
+- (NSDecimalNumber *)starRating {
+    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%ld",(long)_nativeAd.data.score]];
+}
+
+- (NSString *)store {
+    return nil;
+}
+
+- (NSString *)price {
+    return nil;
+}
+
+- (NSString *)advertiser {
+    if (_nativeAd && _nativeAd.data) {
+        return _nativeAd.data.AdTitle;
+    }
+    return nil;
+}
+
+- (NSDictionary<NSString *,id> *)extraAssets {
+    if (_nativeAd) {
+        return @{BUDNativeAdTranslateKey:_nativeAd};
+    }
+    return nil;
+}
+
+- (void)didUntrackView:(UIView *)view {
```

Comment:
@wangyuanyuanyy This was just added, may I ask why it wasn't added before?

Reply:
This is an optional method. I found it when I revised it yesterday, and it feels better if it is added.